ledger/eval.go

```diff
@@ -697,14 +656,100 @@ func (eval *BlockEvaluator) GenerateBlock() (*ValidatedBlock, error) {
 	return &vb, nil
 }
 
+type evalTxValidator struct {
+	txcache          VerifiedTxnCache
+	block            bookkeeping.Block
+	proto            config.ConsensusParams
+	verificationPool execpool.BacklogPool
+
+	ctx      context.Context
+	cf       context.CancelFunc
+	txgroups chan []transactions.SignedTxnWithAD
+	done     chan error
+}
+
+func (tv *evalTxValidator) run() {
+	for txgroup := range tv.txgroups {
+		select {
+		case <-tv.ctx.Done():
+			tv.done <- tv.ctx.Err()
+			tv.cf()
+			close(tv.done)
+			return
+		default:
+		}
+		groupNoAD := make([]transactions.SignedTxn, len(txgroup))
+		for i := range txgroup {
+			groupNoAD[i] = txgroup[i].SignedTxn
+		}
+		ctxs := verify.PrepareContexts(groupNoAD, tv.block.BlockHeader)
+
+		for gi, tx := range txgroup {
+			err := validateTransaction(tx.SignedTxn, tv.block, tv.proto, tv.txcache, ctxs[gi], tv.verificationPool)
+			if err != nil {
+				//logging.Base().Errorf("evalTxValidator reject tx %s", tx.ID())
+				tv.done <- err
+				tv.cf()
+				close(tv.done)
+				return
+			}
+		}
+	}
+	close(tv.done)
+}
+
+func validateTransaction(txn transactions.SignedTxn, block bookkeeping.Block, proto config.ConsensusParams, txcache VerifiedTxnCache, ctx verify.Context, verificationPool execpool.BacklogPool) error {
+	// Transaction valid (not expired)?
+	err := txn.Txn.Alive(block)
+	if err != nil {
+		return err
+	}
+
+	// Well-formed on its own?
+	err = txn.Txn.WellFormed(ctx.CurrSpecAddrs, proto)
+	if err != nil {
+		return fmt.Errorf("transaction %v: malformed: %v", txn.ID(), err)
+	}
+
+	if txcache == nil || !txcache.Verified(txn, ctx.Params) {
+		err = verify.TxnPool(&txn, ctx, verificationPool)
+		if err != nil {
+			return fmt.Errorf("transaction %v: failed to verify: %v", txn.ID(), err)
+		}
+	}
+	return nil
+}
+
+// used by Ledger.Validate() Ledger.AddBlock() Ledger.trackerEvalVerified()(accountUpdates.loadFromDisk())
+//
+// Validate: eval(ctx, blk, true, txcache, executionPool)
+// AddBlock: eval(context.Background(), blk, false, nil, nil)
+// tracker:  eval(context.Background(), blk, false, nil, nil)
 func (l *Ledger) eval(ctx context.Context, blk bookkeeping.Block, validate bool, txcache VerifiedTxnCache, executionPool execpool.BacklogPool) (StateDelta, error) {
```

Comment:
@tsachiherman when this is called from Ledger.Validate() it will use the execution pool passed into Validate() to process the signature checks, same as happened before this refactor.